### PR TITLE
Indexed attestations store

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---

indexed attestation store for slasher

### CODE REVIEW

**Comment by nisdas** [comment]
@shayzluf is this still relevant ?

**Comment by shayzluf** [comment]
Sure, and I think is should be ready. Will check and let you know

File: .bazelrc [hunk 1]
```diff
@@ -6,7 +6,7 @@ test --build_tests_only
 test --test_output=errors
 
 # Fix for rules_docker. See: https://github.com/bazelbuild/rules_docker/issues/842
-build --host_force_python=PY2
+build --host_force_python=PY2 --incompatible_strict_action_env
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply .bazelrc]: was preston suggestion to remove the need for bazel to rebuild when moving from terminal to ide

File: proto/eth/v1alpha1/slasher.proto [hunk 2]
```diff
@@ -29,10 +29,21 @@ service Slasher {
...
     rpc SlashableAttestations(google.protobuf.Empty) returns (stream ethereum.eth.v1alpha1.AttesterSlashing);
 }
+
+message ValidatorIDToIndexedAttestation {
+    repeated uint64 indices = 1 ;
+    bytes data_root = 2;
+    // 96 bytes aggregate signature.
+    bytes signature = 3;
+}
+message ValidatorIDToIndexedAttestationList {
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: shorten it. i hope its descriptive enough

File: proto/eth/v1alpha1/slasher.proto [hunk 3]
```diff
@@ -29,10 +29,21 @@ service Slasher {
...
     rpc SlashableAttestations(google.protobuf.Empty) returns (stream ethereum.eth.v1alpha1.AttesterSlashing);
 }
+
+message ValidatorIDToIdxAtt {
+    repeated uint64 indices = 1 ;
+    bytes data_root = 2;
+    // 96 bytes aggregate signature.
+    bytes signature = 3;
+}
+message ValidatorIDToIdxAttLst {
```
Comment (terencechain) [inline proto/eth/v1alpha1/slasher.proto]: Is this a typo? Do you mean to spell `List` not `Lst`?

Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: it was part of an effort to shorten the var name https://github.com/prysmaticlabs/prysm/pull/3322#discussion_r335874295

File: slasher/db/indexed_attestations.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
+package db
+
+import (
+	"bytes"
+	"reflect"
+	"sort"
+
+	"github.com/prysmaticlabs/go-ssz"
+
```
Comment (terencechain) [inline slasher/db/indexed_attestations.go]: goimports

File: slasher/db/indexed_attestations.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+			}
+		}
+		hasAttestation = false
+		return nil
+	})
+
+	return hasAttestation
+}
+
+// SaveIndexedAttestation accepts epoch  and indexed attestation and writes it to disk.
```
Comment (terencechain) [inline slasher/db/indexed_attestations.go]:
```suggestion
// SaveIndexedAttestation accepts epoch and indexed attestation and writes it to disk.
```

File: slasher/server_test.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+	if len(sr.ProposerSlashing) != 1 {
+		t.Errorf("Should return 1 slashaing proof: %v", sr)
+	}
+	if !proto.Equal(sr.ProposerSlashing[0], want) {
+		t.Errorf("wanted slashing proof: %v got: %v", want, sr.ProposerSlashing[0])
+
+	}
+
```
Comment (terencechain) [inline slasher/server_test.go]:
```suggestion
```

File: slasher/server_test.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+		Header_1:      psr2.BlockHeader,
+		Header_2:      psr.BlockHeader,
+	}
+
+	if len(sr.ProposerSlashing) != 1 {
+		t.Errorf("Should return 1 slashaing proof: %v", sr)
+	}
+	if !proto.Equal(sr.ProposerSlashing[0], want) {
+		t.Errorf("wanted slashing proof: %v got: %v", want, sr.ProposerSlashing[0])
+
```
Comment (terencechain) [inline slasher/server_test.go]:
```suggestion
```

File: slasher/server_test.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	sr, err := slasherServer.IsSlashableBlock(ctx, psr2)
+	if err != nil {
+		t.Errorf("Could not call RPC method: %v", err)
+	}
+
+	if len(sr.ProposerSlashing) != 0 {
+		t.Errorf("Should return 0 slashaing proof: %v", sr)
+	}
+
```
Comment (terencechain) [inline slasher/server_test.go]:
```suggestion
```

File: slasher/server_test.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+	if err != nil {
+		t.Errorf("Could not call RPC method: %v", err)
+	}
+
+	if len(sr.ProposerSlashing) != 1 {
+		t.Errorf("Should return 1 slashaing proof: %v", sr)
+	}
+	if !proto.Equal(sr.ProposerSlashing[0], want) {
+		t.Errorf("wanted slashing proof: %v got: %v", want, sr.ProposerSlashing[0])
+
```
Comment (terencechain) [inline slasher/server_test.go]:
```suggestion
```

File: slasher/server_test.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	sr, err := slasherServer.IsSlashableBlock(ctx, psr)
+	if err != nil {
+		t.Errorf("Could not call RPC method: %v", err)
+	}
+
+	if len(sr.ProposerSlashing) != 0 {
+		t.Errorf("Should return 0 slashaing proof: %v", sr)
+	}
+
```
Comment (terencechain) [inline slasher/server_test.go]:
```suggestion
```

File: proto/eth/v1alpha1/slasher.proto (hunk 3 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: chnaged

File: slasher/server.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
+package slasher
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
```
Comment (terencechain) [inline slasher/server.go]: goimport

**Pull request merged**